include/picongpu/plugins/binning/WriteHist.hpp

```diff
@@ -0,0 +1,171 @@
+
+                /**
+                 * Write the histogram
+                 */
+                ::openPMD::Mesh mesh = iteration.meshes["Binning"];
+
+                // Call the user defined OpenPMD
+                binningData.writeOpenPMDFunctor(series, iteration, mesh);
+
+                mesh.setGeometry(::openPMD::Mesh::Geometry::cartesian);
+                mesh.setDataOrder(::openPMD::Mesh::DataOrder::C);
+
+                std::apply(
+                    [&](auto&... tupleArgs)
+                    {
+                        ((mesh.setAttribute(tupleArgs.label + "_bin_edges", tupleArgs.getBinEdgesSI())), ...);
+                        ((mesh.setAttribute(tupleArgs.label + "_units", tupleArgs.units)), ...);
+                        std::vector<std::string> labelVector;
+                        ((labelVector.emplace_back(tupleArgs.label)), ...);
+                        std::reverse(labelVector.begin(), labelVector.end());
+                        mesh.setAxisLabels(labelVector);
+                    },
+                    binningData.axisTuple); // careful no const tupleArgs
+
+                std::vector<double> gridSpacingVector;
+                std::vector<double> gridOffsetVector;
+
+                for(int i = binningData.getNAxes() - 1; i >= 0; i--)
+                {
+                    gridSpacingVector.emplace_back(1.); // How to deal with non fixed grid spacings?
+                    gridOffsetVector.emplace_back(0.);
+                }
+                mesh.setGridSpacing(gridSpacingVector);
+                mesh.setGridGlobalOffset(gridOffsetVector);
+
+                {
+                    using UD = ::openPMD::UnitDimension;
+                    mesh.setUnitDimension(makeOpenPMDUnitMap(binningData.depositionData.units)); // charge density
+                }
+
+                ::openPMD::MeshRecordComponent record = mesh[::openPMD::RecordComponent::SCALAR];
+
+                /*
+                 * The value represents an aggregation over one cell, so any value is correct for the mesh position.
+                 * Just use the center.
+                 */
+                record.setPosition(std::vector<float>{0.5, 0.5});
+
+                ::openPMD::Offset histOffset;
+                ::openPMD::Extent histExtent;
+                auto bufferExtents = binningData.axisExtentsND;
+
+                /** Z Y X - reverse order */
+                for(int i = binningData.getNAxes() - 1; i >= 0; i--)
+                {
+                    histExtent.emplace_back(static_cast<size_t>(bufferExtents[i]));
+                    histOffset.emplace_back(static_cast<size_t>(0));
+                }
+
+                record.setUnitSI(get_conversion_factor(outputUnits));
+
+                record.resetDataset({::openPMD::determineDatatype<Type>(), histExtent});
+                std::shared_ptr<Type> data(hReducedBuffer.getBasePointer(), [](auto const&) {});
+                record.storeChunk<Type>(data, histOffset, histExtent);
```

Comment:
```suggestion
#if OPENPMDAPI_VERSION_GE(0, 15, 0)
                ::openPMD::UniquePtrWithLambda<Type> data(
                    hReducedBuffer.getBasePointer(),
                    [bfr = std::move(hReducedBuffer)](auto const*) {});
#else
                std::shared_ptr<Type> data(hReducedBuffer.getBasePointer(), [](auto const&) {});
#endif
                record.storeChunk<Type>(std::move(data), histOffset, histExtent);
```

This can help save memory in the BP5 engine when using a somewhat recent installation of openPMD-api. 
I'll need to check the exact version again and the `hReducedBuffer` would then need to be passed into this whole function by move instead of by reference.

If I don't forget it, I'll prepare a commit for this next week.

Reply:
Hm, I get errors from this. Need to look more closely next week.

Reply:
Thanks for looking into it. We will keep this PR the code as it is and your optimization should go into the code a separate optimization.
I assume your optimization should be applied to other parts of PIConGPU too.

Reply:
IMO `std::move(hReducedBuffer)` is not allowed because `hReducedBuffer` is managed by PIConGPU and the plugin is reusing the buffer. If you move the object into the lambda it would mean you are not allowed to touch the buffer anymore.

Reply:
> I assume your optimization should be applied to other parts of PIConGPU too.

No, most other parts of PIConGPU already use `storeChunkSpan()` from `include/picongpu/plugins/common/openPMDVersion.def` which is a better optimization than this one, but would require heavy restructuring here.

Reply:
> IMO `std::move(hReducedBuffer)` is not allowed because `hReducedBuffer` is managed by PIConGPU and the plugin is reusing the buffer. If you move the object into the lambda it would mean you are not allowed to touch the buffer anymore.

Is that so? The way I see it, `Binner::notify()` in `include/picongpu/plugins/binning/Binner.hpp` allocates the buffer and the last use of it is in the call to `histWriter()`. Moving the buffer is the entire point of this optimization, so if we can't do that, then we can leave things as they are.

Reply:
> then we can leave things as they are.

.. we should at least use `storeChunkRaw()` instead of a fake shared pointer though

Reply:
If I read `include/pmacc/memory/buffers/HostBuffer.hpp` correctly, then `HostBuffer` does not implement move semantics? There is nothing that ever overrides `pointer` and `ownPointer`, so after a move you get a double free.
(If we don't want to implement those operations, it would probably be good to explicitly `delete` them, to avoid others falling into the same problem)

Reply:
[This commit](https://github.com/ComputationalRadiationPhysics/picongpu/commit/fe11baebaa840d7dd1a280e10b93f3c8389843ed) has the full suggestion for adding this optimization. Alternatively to the workaround there (putting the HostBuffer into a unique_ptr), one could also add move semantics to HostBuffer.